fake_tensor: Handle op errors more gracefully

c00w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147049

if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @StrongerXi

c00w committed:
fake_tensor: Handle op errors more gracefully

if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

[ghstack-poisoned]

jansel reviewed (changes requested):
failing tests?

c00w commented:
@jansel  - Will do, but wanted to make sure overall direction here makes sense (i.e. for RuntimeError, we should graph break, and let eager mode raise the error, rather than explicitly throwing it from within dynamo). My read is that you agree this direction makes sense, but wanted to be explicit :).

jansel commented:
Graph breaking on things that will error in AOTAutograd seems fine to me.

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w commented:
I believe all errors are fixed - Note that the test case for test_binary_ufuncs makes the error message look worse than it is. You actually get the original error shown as causing the new error, but AFAIK there isn't a way to test for that in python with assertRaises.

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w commented:
Not all errors are fixed, looking.

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

jansel reviewed (changes requested):
re-request review once tests are passing

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

bdhirsh reviewed (changes requested):
more conversation in the link, but this is morally the same as setting `suppress_errors=True` for all FakeTensor errors in PT2, which has the risk of hiding problems: https://fb.workplace.com/groups/1286739428954016/posts/1432091381085486/?comment_id=1432931107668180

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
None on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]

c00w committed:
Update on "fake_tensor: Handle op errors more gracefully"


if we have a operator error (i.e. incompatible dimensions etc... from
torch._check) within a faketensor, then it fails with
torch._dynamo.exc.TorchRuntimeError rather than gracefully falling back to be
unimplemnted and letting eager mode fail

Since fake_tensor had no dependency on dynamo, I did not add one, and instead
relied up existing fake tensor unsupported exceptions. Let me know if there is
a preference to instead use ObservedExceptions here.

I've added a specific test case which triggers a torch._check failure whose
stack trace lines up with errors I've observed in production

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames StrongerXi

[ghstack-poisoned]
